### PR TITLE
store/tikv: wait task channel in Close

### PR DESCRIPTION
**Pull request by coocood**
Makes test more stable.

Close distsql result and partial result in executor.

### CODE REVIEW

File: executor/distsql.go [hunk 1]
```diff
@@ -888,6 +891,7 @@ func (e *XSelectTableExec) Next() (*Row, error) {
 		}
 		if rowData == nil {
 			// Finish the current partial result and get the next one.
+			e.partialResult.Close()
```
Comment (shenli) [inline executor/distsql.go]: This will fix memory leak?

Comment (coocood) [inline executor/distsql.go]: The partial result should be closed, may fix the potential memory leak.

Comment (coocood) [inline executor/distsql.go]: But in this case, the result has been consumed already, there should not have memory leak.

**Comment by shenli** [comment]
LGTM

File: executor/distsql.go (hunk 1 above)
Comment (tiancaiamao) [inline executor/distsql.go]: 1. partialResult.Close() is a null operation, in distsql/distsql.go

```
// Close closes the sub result.
func (pr *partialResult) Close() error {
	return nil
}
```

2. I'm afraid Close is called many times.

File: store/tikv/coprocessor.go [hunk 2]
```diff
@@ -488,6 +488,7 @@ func (it *copIterator) handleRegionErrorTask(bo *Backoffer, task *copTask) []cop
 
 func (it *copIterator) Close() error {
 	close(it.finished)
+	it.wg.Wait()
```
Comment (tiancaiamao) [inline store/tikv/coprocessor.go]: Why need wait here? It's already used in `run()` function.

Comment (coocood) [inline store/tikv/coprocessor.go]: To make sure the worker goroutines have exited.

File: executor/distsql.go [hunk 3]
```diff
@@ -533,10 +532,12 @@ func (e *XSelectIndexExec) addWorker(workCh chan *lookupTableTask, concurrency *
...
 
 func (e *XSelectIndexExec) fetchHandles(idxResult distsql.SelectResult, ch chan<- *lookupTableTask) {
-	defer close(ch)
-
 	workCh := make(chan *lookupTableTask, 1)
-	defer close(workCh)
+	defer func() {
+		close(ch)
+		close(workCh)
+		idxResult.Close()
```
Comment (tiancaiamao) [inline executor/distsql.go]: move it to `executeTask` and write like this?

```
	tblResult, err := e.doTableRequest(task.handles)
	if err != nil {
		return errors.Trace(err)
	}
    defer tblResult.Close()
```

Comment (coocood) [inline executor/distsql.go]: It's index result, not table result.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**